src/ripple/app/tx/impl/Transactor.cpp

```diff
@@ -448,6 +465,31 @@ Transactor::apply()
     return doApply();
 }
 
+NotTEC
+Transactor::checkTemporalTxValidity(PreclaimContext const& ctx)
+{
+    using d = NetClock::duration;
+    using tp = NetClock::time_point;
+
+    if (ctx.view.rules().enabled(featureTemporalTxValidity))
+    {
+        // We check time against the close time of the parent ledger,
+        // because that is the time that the network, as a whole,
+        // agrees on.
+        auto const pct = ctx.view.parentCloseTime();
+
+        if (auto const cutoff = ctx.tx[~sfNotValidBefore];
+            cutoff && (pct < tp{d{*cutoff}}))
+            return tefTOO_EARLY;
```

Comment:
I feel `tefTOO_EARLY` and `tefTOO_LATE` are slightly different since a tx that is too early could be retried without modification and a tx that is too late cannot be retried. So could `tefTOO_EARLY` be `terTOO_EARLY`?
(Sorry I am not sure the consequences if we change, such as its behavior in `TxQ`. I could run some tests later if needed.)

Reply:
> @scottschurr wrote:
> There may be some value in supplying both a terTOO_EARLY and a tefTOO_EARLY. The tefTOO_EARLY code is the more general one. If the transaction is a day, an hour, or even 5 minutes early, then everyone is better off if the transaction simply fails. The number of retries that the server is willing perform will not make a transaction that is 5 minutes early succeed.
>
> However, if the transaction is 10 seconds early or less, then it makes sense for the server to retry the transaction a few times. This is the case where terTOO_EARLY makes sense.
>
> The question we have to ask is whether it's worth adding code to support the "just a little bit early" case. I could go either way, but I have a slight bias that the added complexity does not bring enough usability value.

Rhetorical question: does the distinction between "ter" and "tef" classes exist for the sake of the user, or for the sake of the server? Is the purpose to make it easier for the user to know how to process the outcome of the transaction? Or is the purpose to direct the server in how to handle the failure?

If the answer is for the sake of the user, then this should be a `ter` code, to signal to the user, "This transaction, unmodified, could become valid later."

If the answer is for the sake of the server, then making it a `tef` code so that the server does not need to retry it fruitlessly makes sense.

Reply:
It turns out that `tem` and `ter` codes are handled somewhat differently inside the server.  So the codes are intended for both parties.  But which one you select has consequences for how long the server holds onto the transaction before giving up.  The server will work harder at pushing a `ter` into the ledger, or at least that was true the last time I looked.  I have to admit, these things evolve over time...